Initial version of mypy.ini file

waseem18 opened the pull request:
Related to #231 

This PR adds a new file named `mypy.ini` which is a configuration file of `mypy`.
This is an initial version of `mypy.ini`, as we proceed on type hinting the code we may add more rules.

waseem18 commented:
@brainwane With this commit - I'm starting to work on #231 
As `mypy.ini` is the heart of `mypy` configuration, could you please review this PR and let me know if I missed anything. I took inspiration from `zulip` in adding the rules to `mypy.ini`.

sigmavirus24 commented on mypy.ini:
```diff
@@ -0,0 +1,8 @@
+[mypy]
+check_untyped_defs = True
+disallow_any_generics = True
```
Is it wise to start with disallowing generics? Given some of our code is meant to become a library, I'm wondering if we'll be using generics to provide a sensible interface.

waseem18 replied:
Thanks for the review @sigmavirus24 . I'll change that.

waseem18 commented:
@sigmavirus24 I've got a question, I'll be glad if you can help me with them.

In `tox.ini` file we see 

`[testenv:pep8]`
`basepython = python2.7`

As `basepython` is set to `2.7` whenever we run `tox -e pep8` we'd get `E999 SyntaxError: invalid syntax`. We should probably change it to `3.6` or have multiple versions?

sigmavirus24 commented:
@waseem18 twine needs to work on python 2.7 so if you're seeing syntax error on Python 2.7 that means twine will be unable to run on 2.7. That would imply to me that you're adding the hints directly to the files when the discussion in the related issue agreed upon using stub files.

waseem18 commented:
@sigmavirus24 I've added `_stubs` folder and type hinted a couple of files. As we're placing the type hints in a separate file within a folder named `_stubs`, the way we would need to run `mypy` would be through test files. 

For example, `mypy tests/test_main.py` would check `main.pyi` file. 

In the two files `_stubs/__main.pyi__` and `_stubs/cli.py` I've added those methods which are mentioned in the respective test files. In other words I didn't type hint the internal methods (methods which generally start with `_` or are not being called from test file). 

Could you please provide me some feedback on this?

Thanks

brainwane commented:
@waseem18 could you also look at #357?

bhrutledge commented:
Continuing this work in #466.

bhrutledge commented:
@di and @sigmavirus24 I believe this can be closed; if there's still interest, I'm happy to continue the work in #469.
